nifi-nar-bundles/nifi-rules-bundle/nifi-rules-processors/src/main/java/org/apache/nifi/rules/processors/RulesRecordProcessor.java

```diff
@@ -0,0 +1,244 @@
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.annotation.lifecycle.OnScheduled;
+import org.apache.nifi.avro.AvroTypeUtil;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.flowfile.FlowFile;
+import org.apache.nifi.flowfile.attributes.CoreAttributes;
+import org.apache.nifi.logging.ComponentLog;
+import org.apache.nifi.processor.AbstractProcessor;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.rules.Action;
+import org.apache.nifi.rules.engine.RulesEngineService;
+import org.apache.nifi.serialization.RecordReader;
+import org.apache.nifi.serialization.RecordReaderFactory;
+import org.apache.nifi.serialization.RecordSetWriter;
+import org.apache.nifi.serialization.RecordSetWriterFactory;
+import org.apache.nifi.serialization.WriteResult;
+import org.apache.nifi.serialization.record.ListRecordSet;
+import org.apache.nifi.serialization.record.MapRecord;
+import org.apache.nifi.serialization.record.Record;
+import org.apache.nifi.serialization.record.RecordSchema;
+import org.apache.nifi.serialization.record.RecordSet;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@EventDriven
+@SideEffectFree
+@SupportsBatching
+@Tags({"record", "rules", "action", "rules engine"})
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@WritesAttributes({
+        @WritesAttribute(attribute = "record.count", description = "The number of records in an outgoing FlowFile"),
+        @WritesAttribute(attribute = "mime.type", description = "The MIME Type that the configured Record Writer indicates is appropriate"),
+})
+@CapabilityDescription("Submits record values to a rules engine and returns actions determined by the engine as records ")
+public class RulesRecordProcessor extends AbstractProcessor {
+
+    static final PropertyDescriptor FACTS_RECORD_READER = new PropertyDescriptor.Builder()
+            .name("fact-record-reader")
+            .displayName("Fact Record Reader")
+            .description("Specifies the Controller Service to use for parsing incoming fact data and determining the data's schema.")
+            .identifiesControllerService(RecordReaderFactory.class)
+            .required(true)
+            .build();
+
+    static final PropertyDescriptor ACTION_RECORD_WRITER = new PropertyDescriptor.Builder()
+            .name("action-record-writer")
+            .displayName("Action Record Writer")
+            .description("Specifies the Controller Service to use for writing out action records")
```

Comment:
Perhaps we should mention here that Inherit Schema on the RecordWriter will allow it to use the generated/static schema provided by the processor. The user could still override it (to include/exclude fields) but might be helpful to point out that Inherit Schema wouldn't inherit the reader schema, but the provided one.

Reply:
Will do